sources/kafka/helpers.py

```diff
@@ -0,0 +1,145 @@
+    """Initiate a Kafka consumer with the given credentials.
+
+    Args:
+        credentials (dict): Kafka consumer credentials.
+        group_id (str): Kafka consumer group.
+
+    Returns:
+        confluent_kafka.Consumer: A ready to go Kafka consumer.
+    """
+    config = {
+        "bootstrap.servers": credentials["bootstrap"]["servers"],
+        "group.id": group_id,
+        "security.protocol": credentials["security"]["protocol"],
+        "sasl.mechanisms": credentials["sasl"]["mechanisms"],
+        "sasl.username": credentials["sasl"]["username"],
+        "sasl.password": credentials["sasl"]["password"],
+    }
+    return Consumer(config)
+
+
+class OffsetTracker(dict):
+    """Object to control offsets of the given topics.
+
+    Tracks all the partitions of the given topics with two params:
+    current offset and maximum offset (partition length).
+
+    Args:
+        consumer (confluent_kafka.Consumer): Kafka consumer.
+        topic_names (list): Names of topics to track.
+    """
+
+    def __init__(self, consumer: Consumer, topic_names: list):
+        super().__init__()
+
+        self._consumer = consumer
+        self._topics = self._read_topics(topic_names)
+
+        # read/init current offsets
+        self._cur_offsets = dlt.current.resource_state().setdefault(
+            "offsets", {t_name: {} for t_name in topic_names}
+        )
+
+        self._init_partition_offsets()
+
+    def _read_topics(self, topic_names):
+        """Read the given topics metadata from Kafka.
+
+        Reads all the topics at once, instead of requesting
+        each in a separate call. Returns only those needed.
+
+        Args:
+            topic_names (list): Names of topics to be read.
+
+        Returns:
+            dict: Metadata of the given topics.
+        """
+        tracked_topics = {}
+        topics = self._consumer.list_topics().topics
+
+        for t_name in topic_names:
+            tracked_topics[t_name] = topics[t_name]
+
+        return tracked_topics
+
+    def _init_partition_offsets(self):
+        """Designate current and maximum offsets for every partition.
+
+        Current offsets are read from the state, if present. Set equal
+        to the partition beginning otherwise.
+        """
+        all_parts = []
+        for t_name, topic in self._topics.items():
+            self[t_name] = {}
+
+            # init all the topic partitions from the partitions' metadata
+            parts = [TopicPartition(t_name, part) for part in topic.partitions]
+
+            # designate current and maximum offsets for every partition
+            for i, part in enumerate(parts):
+                cur_offset = self._cur_offsets[t_name].get(
+                    str(part.partition), OFFSET_BEGINNING
+                )
+
+                self[t_name][str(part.partition)] = {
+                    "cur": cur_offset,
+                    "max": self._consumer.get_watermark_offsets(part)[1],
```

Comment:
Yeah, I know, but I don't see how to do it in one request, even `AdminClient` does it per one partition. There are several feature requests in the Confluent Kafka repos for methods, which could give all the watermarks, however, they are all open. I think Kafka is just not intended for operations like this, that's why they don't rush to implement it.

Reply:
OK! but here you can get offset for timestamps `ts_offsets = self._consumer.offsets_for_times(parts)` so you could do the same to get offset for `now`.
but maybe there is a reason not to do that :) anyway - we can live with the thing below

Reply:
Actually, I thought about the same, I even wrote you a message, but then deleted it :sweat_smile: 

Offset is a cursor at which the **user** is reading now. It's not the same as the **topic's** cursor for new messages insert. Messages are added to the end of the topic, but the user's offset may be somewhere in the middle. And we need both - to understand where to start reading (incremental), and when to stop (not to get into endless loop).

And when we're talking about getting offsets for timestamps, it's about user's offset, without the topic length. Also, there is quite strange trait of behavior in it:
```
If the provided timestamp exceeds that of the last message
in the partition, a value of -1 will be returned.
```
Seems to me we can't use `now` or any values close to it, because it'll give `-1` in most of the cases. `-1` means `no offset available for partition [...] happens at the very first read, or the group's offset expiration/deletion`.

I must say it's a very strange behavior by my opinion, but it's what it is. So, we can start reading from a particular timestamp, but we still don't know when to stop, and we'll need to somehow process situations, when our timestamp is after the last message.

Reply:
uh! this sucks. we must add checks against that in our code if indeed
`ts_offsets = self._consumer.offsets_for_times(parts)`
may create offset of -1 when our timestamp is too new, we must fallback to max time from watermark when creating curr_offset

also you must update the pipeline state at the end `_init_partition_offsets` with the current offsets. so in the case above we use the stored watermark as starting point when pipeline is run ie. an hour later

Reply:
this is also important test case
- send a message
- consume with start timestamp to now. you should get -1 as the offset
- the code should fallback to watermark and store in the state as current offset. it returns no messages
- then you send a message
- run the pipeline again and the second message should be returned

Reply:
Done